test/upgrade-test.go

```diff
@@ -0,0 +1,565 @@
+	}
+	return migratorBase
+}
+
+// Check drift with docker run migrator, use latest migrator version
+func migratorDrift(version version, ctx context.Context, db string, migratorArgs ...string) error {
+	fmt.Println("Checking for drift with docker run migrator ... ")
+	// Get latest migrator version
+	tag, err := getLatestMigrator(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to select latest migrator version: %s", err)
+	}
+	// Construct the docker run command for migrator
+	migratorBase := migratorBaseString(tag, ctx)
+	migratorCmd := append(migratorBase, "drift", "--db="+db, "--version="+version.vNum())
+	migratorCmd = append(migratorCmd, migratorArgs...)
+	// Run the constructed docker run migrator command
+	err = run.Cmd(ctx, migratorCmd...).Run().Stream(os.Stdout)
+	if err != nil {
+		return fmt.Errorf("drift check failed for %s version %s: %s", db, version.vNum(), err)
+	} else {
+		fmt.Println("No schema drift found in", db)
+		return nil
+	}
+}
+
+// Check drift with docker run migrator
+func migratorUpgrade(vFrom version, vTo version, ctx context.Context, migratorArgs ...string) error {
+	fmt.Printf("\nPerforming MVU from %s to %s ... ", vFrom.vNum(), vTo.vNum())
+	// Construct the docker run command for migrator
+	migratorBase := migratorBaseString(vTo, ctx)
+	migratorCmd := append(migratorBase, "upgrade", "--from="+vFrom.vNum(), "--to="+vTo.vNum())
+	migratorCmd = append(migratorCmd, migratorArgs...)
+	// Run the constructed docker run migrator command
+	err := run.Cmd(ctx, migratorCmd...).Run().Stream(os.Stdout)
+	if err != nil {
+		return fmt.Errorf("multiversion upgrade failed from %s to %s: %s", vFrom.vNum(), vTo.vNum(), err)
+	}
+	return nil
+}
+
+// checkout a version tag
+func gitCheckoutVersion(version version, ctx context.Context) error {
+	fmt.Println("Checking out version " + version.vNum())
+	err := run.Cmd(ctx, "git", "checkout", version.vNum()).Run().Stream(os.Stdout)
+	if err != nil {
+		return fmt.Errorf("failed to checkout version %s: %s", version.vNum(), err)
+	}
+	return nil
+}
+
+// Docker commands
+
+// Prune docker volumes
+func dockerPrune(ctx context.Context) error {
+	fmt.Println("Pruning docker volumes...")
+	err := run.Cmd(ctx, "docker", "volume", "prune", "-a", "-f").Run().Stream(os.Stdout)
+	if err != nil {
+		return fmt.Errorf("failed to prune docker volumes: %s", err)
+	}
+	return nil
+}
+
+// composeUpTimeout creates a timed context to handle for the failure of migrator during `docker-compose up`
+// Migrator "up" in docker-compose can crash loop if it fails, with the migrator continually trying to run the "up" command
+// and the frontend waiting to initialize after migrator's successful completion.
+//
+// This command has a 80 sec timeout and will check the migrator logs to display the failure if the timeout is reached.
+func composeUpTimeout(ctx context.Context, images ...string) error {
+	timeout := 80 * time.Second
+	tCtx, cancel := context.WithTimeout(context.Background(), timeout)
```

Comment:
This one I don't want it to use the parent context, I want a new context here so that the timeout isn't attached to the context inherited by main. Am I thinking about that right?

Reply:
In this case, if you cancel `tCtx` it won't effect the parent context so you should be fine.

Reply:
The code for `context.WithTimeout` shows that if you try to put a timeout on a context that already has a timeout, it will use the earliest deadline of the two, which will cancel both.

Reply:
Is this mostly a style thing? I went ahead and changed it but it seems like I could do it either way if I don't mind this having no connection to the parent context. Based on what you're saying JH it seems like if you plan on having multiple timeouts its safest to declare new contexts for a function with a timeout?

Reply:
Alright yeah I ended up just using ctx inherited from main and its nbd